### ISSUE TITLE
GUI window closure shoudl trigger first the proper deinitialization of modules

### ISSUE THREAD

**Issue by seb5g** [PyMoDAQ/PyMoDAQ#374]
When clicking on the close window X button. The close event should be intercepted to first trigger the proper closure of all activated modules (dashoboard, daq_move and others)

**Comment by bernhardlang** [comment]
Though this item dates already a bit, here a suggestion, in case that hasn't been implemented yet: install an event filter on the main window. For instance

class CloseEventFilter(QObject):

    def __init__(self, application):
        """application: CustomApp/CustomExt/Dashboard/similar"""
        super().__init__()
        self.close_enabled = True
        self.application = application

    def eventFilter(self, obj, event):
        if event.type() == QEvent.Close:
            if not self.close_enabled:
                event.ignore()
                return True
            self.application.quit_function()
        return super().eventFilter(obj, event)

and in the __init__ function of the CustomApp/CustomExt/... 

        self.close_event_filter = CloseEventFilter(self)
        self.mainwindow.installEventFilter(self.close_event_filter)

The filter object may either intercept closing completely (e.g. prohibiting unwanted application death by accidentally closing the main window during measurement or when yet unsaved data is around). Or it inserts the application's quit_function which may perform whatever shutdown operations are needed. I tested this approach on a custom app / extension, calling self.detector.quit_fun() (that's a DAQ_Detector) from inside my application's quit_function. Sometimes I get 

python3.11/site-packages/pymodaq/control_modules/daq_viewer.py", line 1244, in queue_command
    self.status_sig.emit(ThreadCommand(command.command, [status, 'log']))
    ^^^^^^^^^^^^^^^
RuntimeError: wrapped C/C++ object of type DAQ_Detector has been deleted

Probably, one would first have to shut down the non-GUI threads and wait until they have joined to solve that issue. In case that there are multiple QMainWindows around (CustomExt + Dashboard not closed), one would probably have to react on QApplication.lastWindowClosed before the latter is forwarded to QApplication.quit).


Alternative approach: inherit QMainWindow.closeEvent

class CustomMain(QMainWindow):

    def __init__(self):
        QMainWindow.__init__(self)
        self.close_enabled = True

    def closeEvent(self, event):
        if self.close_enabled:
            self.application.clean_up()
            QMainWindow.closeEvent(self, event)

This works with CustomApp because the mainwindow is created in the function main() inside the custom app's file (tested). With a CustomExt, it is not possible directly because of line 363 in dashboard.py

self.extension_windows.append(QtWidgets.QMainWindow())

This approach would ask for some kind of factory solution, calling here a function like

    def make_mainwindow(self, kwargs):
        return QMainWindow()

which can be overwritten as needed in CustomApps / Extensions which need additional functions in their mainwindow.

**Comment by seb5g** [comment]
yes I think your second approach is the good one. We don't really need a factory but probably just a inherited MainWindow class dedicated and declared somewhere in the pymodaq_gui menu (maybe somewhere near the CustomApp module)

**Comment by seb5g** [comment]
@bernhardlang would you care implementing it? @malik-irain could you comment on the proposed solution ?

**Comment by bernhardlang** [comment]
My programming assistant promised me to get to it before the week with the three Wednesdays :-) 
A couple follow-up questions:
  * CustomApp takes the dockarea as argument. Therefore, InheritedMainWindow as parent of the latter should probably be created beforehand, doesn't it? InheritedMainWindow in turn has to know about the CustomApp to be able to trigger the shutdown -> sort of chicken and egg. If the dockarea is guaranteed to have an InheritedMainWindow as parent when CustomApp is created, telling InheritedMainWindow (inferred from dockarea.parent) about CustomApp inside he latter's __init__()  is probably the most straight forward solution. 
  * What has to be called to properly shut down the machinery? Does this have to take care of joining multiple threads as well? 
  * Should the shutdown operation be extensible to add further CustomApp/Ext defined stuff when needed?
  * In the pymodaq_plugins_teaching / beam profiler example, the quit action is connected to BeamProfiler(gutils.CustomApp).quit_app(). Should a generic function like that be added to CustomApp, taking care of the shutdown?
  * Should  the inherited QMainWindow be extensible itself?
  * Shall the possibility to completely intercept closing the window be kept?
  * What's about the case where more than one QMainWindow is around, e.g. the dashboard and an extension? Who is responsible for shutting down what? Probably the dashboard for all unless its window has been closed before the extensions window? The latter possibility would mean intercepting at the level QApplication.lastWindowClosed as well or even in first place. However, that is emitted (as well as QCoreApplication.aboutToQuit) __after__ closing has happened. Any part of PyMoDAQ's machinery tied to widgets and windows may already have gone without proper shutdown at that time.
  * A quit action typically being present in the file menu and the taskbar should be kept in the loop as well.

**Comment by malik-irain** [comment]
Close button in system bar is deactivated in extensions, now the main dashboard close trigger should be intercepted

**Comment by bernhardlang** [comment]
I got an experimental version running now (see bernhardlang/pymodaq, bernhardlang/pymodaq_gui and bernhardlang/pymodaq_plugins_readspectro [invoke the app with the cmd line switch --simulate]). In my opinion the main issue is that one should avoid shutting down the machinery while some detectors / viewers are still grabbing data. The cleanest way to do so is probably to inhibit closing while grabbing. That's how I implemented it. Works fine when dealing with a standalone daq_viewer or a single grabbing device. However, with multiple devices, who's going to keep track of the activities? Doing it case by case in a custom app is error prone (and tedious if you've got to maintain several of them). Doing it on a generic level risks ending up in an "usine à gaz".

Furthermore, there's an issue in reaching across threads which may or may not be appropriate, depending on internals of PyMoDAQ I don't oversee. I guess we should have a discussion about that in Montpellier.
